### PR TITLE
[KVM improve logs in migrate VM process

### PR DESCRIPTION
**Pull request by SadiJr**
### Description
This PR aims to improve the logs in migrate VM process, when using KVM as hypervisor, to facilitate future troubleshooting. We faced a situation that the lack of logs in this process made the troubleshooting pretty hard. Therefore, we are proposing to improve it.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

### How Has This Been Tested?
It's tested in a local lab:
1. I created a new VM;
2. I migrated this VM, with its volumes, to another host;
3. I readed the logs to see if they were improved.
4. Before, the logs did not have enough information to perform an efficient troubleshooting;
5. Now, with improved logs, the troubleshooting will be easier.

### CODE REVIEW

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 1]
```diff
@@ -4500,12 +4500,15 @@ public List<Ternary<String, Boolean, String>> cleanVMSnapshotMetadata(Domain dm)
         s_logger.debug("Cleaning the metadata of vm snapshots of vm " + dm.getName());
         List<Ternary<String, Boolean, String>> vmsnapshots = new ArrayList<Ternary<String, Boolean, String>>();
         if (dm.snapshotNum() == 0) {
+            s_logger.debug(String.format("VM [%s] does not have any snapshots. Skipping cleanup of snapshots for this VM.", dm.getName()));
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:4503]:
```suggestion
            if (s_logger.isDebugEnabled()) {
                s_logger.debug(String.format("VM [%s] does not have any snapshots. Skipping cleanup of snapshots for this VM.", dm.getName()));
            }
```

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 2]
```diff
@@ -4500,12 +4500,15 @@ public List<Ternary<String, Boolean, String>> cleanVMSnapshotMetadata(Domain dm)
...
             return vmsnapshots;
         }
         String currentSnapshotName = null;
         try {
             DomainSnapshot snapshotCurrent = dm.snapshotCurrent();
             String snapshotXML = snapshotCurrent.getXMLDesc();
+            s_logger.debug(String.format("Current snapshot of VM [%s] has the following XML: [%s].", dm.getName(), snapshotXML));
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:4510]:
```suggestion
            if (s_logger.isDebugEnabled()) {
                s_logger.debug(String.format("Current snapshot of VM [%s] has the following XML: [%s].", dm.getName(), snapshotXML));
            }
```

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 3]
```diff
@@ -4517,25 +4520,23 @@ public List<Ternary<String, Boolean, String>> cleanVMSnapshotMetadata(Domain dm)
...
         String[] snapshotNames = dm.snapshotListNames();
         Arrays.sort(snapshotNames);
+        s_logger.debug(String.format("Found [%s] snapshots in VM [%s] to clean.", snapshotNames.length, dm.getName()));
         for (String snapshotName: snapshotNames) {
             DomainSnapshot snapshot = dm.snapshotLookupByName(snapshotName);
             Boolean isCurrent = (currentSnapshotName != null && currentSnapshotName.equals(snapshotName)) ? true: false;
             vmsnapshots.add(new Ternary<String, Boolean, String>(snapshotName, isCurrent, snapshot.getXMLDesc()));
         }
         for (String snapshotName: snapshotNames) {
+            s_logger.debug(String.format("Cleaning snapshot [%s] of VM [%s] metadata.", snapshotNames, dm.getName()));
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:4539]:
```suggestion
            if (s_logger.isDebugEnabled()) {
                s_logger.debug(String.format("Cleaning snapshot [%s] of VM [%s] metadata.", snapshotNames, dm.getName()));
            }
```

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 4]
```diff
@@ -102,6 +102,7 @@ public Answer execute(final MigrateCommand command, final LibvirtComputingResour
         final Map<String, Boolean> vlanToPersistenceMap = command.getVlanToPersistenceMap();
         final String destinationUri = createMigrationURI(command.getDestinationIp(), libvirtComputingResource);
         final List<MigrateDiskInfo> migrateDiskInfoList = command.getMigrateDiskInfoList();
+        s_logger.debug(String.format("Trying to migrate VM [%s] to destination host: [%s].", vmName, destinationUri));
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java:105]:
```suggestion
        if (s_logger.isDebugEnabled()) {
            s_logger.debug(String.format("Trying to migrate VM [%s] to destination host: [%s].", vmName, destinationUri));
        }
```

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 5]
```diff
@@ -122,6 +123,7 @@ public Answer execute(final MigrateCommand command, final LibvirtComputingResour
             ifaces = libvirtComputingResource.getInterfaces(conn, vmName);
             disks = libvirtComputingResource.getDisks(conn, vmName);
 
+            s_logger.debug(String.format("Found domain with name [%s]. Starting VM migration to host [%s].", vmName, destinationUri));
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java:126]:
```suggestion
        if (s_logger.isDebugEnabled()) {
            s_logger.debug(String.format("Found domain with name [%s]. Starting VM migration to host [%s].", vmName, destinationUri));
        }
```

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 6]
```diff
@@ -148,13 +150,16 @@ Use VIR_DOMAIN_XML_SECURE (value = 1) prior to v1.0.0.
 
             final String target = command.getDestinationIp();
             xmlDesc = dm.getXMLDesc(xmlFlag);
-            xmlDesc = replaceIpForVNCInDescFile(xmlDesc, target);
+            s_logger.debug(String.format("VM [%s] with XML configuration [%s] will be migrated to host [%s].", vmName, xmlDesc, target));
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java:153]:
```suggestion
        if (s_logger.isDebugEnabled()) {
            s_logger.debug(String.format("VM [%s] with XML configuration [%s] will be migrated to host [%s].", vmName, xmlDesc, target));
        }
```

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 7]
```diff
@@ -148,13 +150,16 @@ Use VIR_DOMAIN_XML_SECURE (value = 1) prior to v1.0.0.
...
+
+            xmlDesc = replaceIpForVNCInDescFile(xmlDesc, target, vmName);
 
             String oldIsoVolumePath = getOldVolumePath(disks, vmName);
             String newIsoVolumePath = getNewVolumePathIfDatastoreHasChanged(libvirtComputingResource, conn, to);
             if (newIsoVolumePath != null && !newIsoVolumePath.equals(oldIsoVolumePath)) {
                 s_logger.debug(String.format("Editing mount path of iso from %s to %s", oldIsoVolumePath, newIsoVolumePath));
                 xmlDesc = replaceDiskSourceFile(xmlDesc, newIsoVolumePath, vmName);
+                s_logger.debug(String.format("Replaced disk mount point [%s] with [%s] in VM [%s] XML configuration. New XML configuration is [%s].", oldIsoVolumePath, newIsoVolumePath, vmName, xmlDesc));
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java:162]:
```suggestion
            if (s_logger.isDebugEnabled()) {
                s_logger.debug(String.format("Replaced disk mount point [%s] with [%s] in VM [%s] XML configuration. New XML configuration is [%s].", oldIsoVolumePath, newIsoVolumePath, vmName, xmlDesc));
            }
```

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 8]
```diff
@@ -174,12 +179,16 @@ Use VIR_DOMAIN_XML_SECURE (value = 1) prior to v1.0.0.
             final boolean migrateStorageManaged = command.isMigrateStorageManaged();
 
             if (migrateStorage) {
+                s_logger.debug(String.format("Changing VM [%s] volumes during migration to host: [%s].", vmName, target));
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java:182]:
```suggestion
            if (s_logger.isTraceEnabled()) {
                s_logger.trace(String.format("Changing VM [%s] volumes during migration to host: [%s].", vmName, target));
            }
```

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 9]
```diff
@@ -174,12 +179,16 @@ Use VIR_DOMAIN_XML_SECURE (value = 1) prior to v1.0.0.
...
                 xmlDesc = replaceStorage(xmlDesc, mapMigrateStorage, migrateStorageManaged);
+                s_logger.debug(String.format("Changed VM [%s] XML configuration of used storage. New XML configuration is [%s].", vmName, xmlDesc));
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java:184]:
```suggestion
            if (s_logger.isDebugEnabled()) {
                s_logger.debug(String.format("Changed VM [%s] XML configuration of used storage. New XML configuration is [%s].", vmName, xmlDesc));
            }
```

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 10]
```diff
@@ -174,12 +179,16 @@ Use VIR_DOMAIN_XML_SECURE (value = 1) prior to v1.0.0.
...
             }
 
             Map<String, DpdkTO> dpdkPortsMapping = command.getDpdkInterfaceMapping();
             if (MapUtils.isNotEmpty(dpdkPortsMapping)) {
+                s_logger.debug(String.format("Changing VM [%s] DPDK interfaces during migration to host: [%s].", vmName, target));
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java:189]:
```suggestion
            if (s_logger.isTraceEnabled()) {
                s_logger.trace(String.format("Changing VM [%s] DPDK interfaces during migration to host: [%s].", vmName, target));
            }
```

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 11]
```diff
@@ -174,12 +179,16 @@ Use VIR_DOMAIN_XML_SECURE (value = 1) prior to v1.0.0.
...
                 xmlDesc = replaceDpdkInterfaces(xmlDesc, dpdkPortsMapping);
+                s_logger.debug(String.format("Changed VM [%s] XML configuration of DPDK interfaces. New XML configuration is [%s].", vmName, xmlDesc));
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java:191]:
```suggestion
            if (s_logger.isDebugEnabled()) {
                s_logger.debug(String.format("Changed VM [%s] XML configuration of DPDK interfaces. New XML configuration is [%s].", vmName, xmlDesc));
            }
```

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 12]
```diff
@@ -262,16 +271,17 @@ Use VIR_DOMAIN_XML_SECURE (value = 1) prior to v1.0.0.
                     }
                 }
             }
-            s_logger.info("Migration thread for " + vmName + " is done");
+            s_logger.info(String.format("Migration thread of VM [%s] finished.", vmName));
 
             destDomain = migrateThread.get(AgentPropertiesFileHandler.getPropertyValue(AgentProperties.VM_MIGRATE_DOMAIN_RETRIEVE_TIMEOUT), TimeUnit.SECONDS);
 
             if (destDomain != null) {
+                s_logger.debug(String.format("Cleaning the disks of VM [%s] in the source pool after VM migration finished.", vmName));
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java:279]:
```suggestion
            if (s_logger.isDebugEnabled()) {
                s_logger.debug(String.format("Cleaning the disks of VM [%s] in the source pool after VM migration finished.", vmName));
            }
```

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 13]
```diff
@@ -313,8 +323,7 @@ Use VIR_DOMAIN_XML_SECURE (value = 1) prior to v1.0.0.
             }
         }
 
-        if (result != null) {
-        } else {
+        if (result == null) {
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java:326]: :+1:

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 14]
```diff
@@ -451,15 +460,17 @@ protected MigrateDiskInfo searchDiskDefOnMigrateDiskInfoList(List<MigrateDiskInf
...
         final int begin = xmlDesc.indexOf(GRAPHICS_ELEM_START);
         if (begin >= 0) {
             final int end = xmlDesc.lastIndexOf(GRAPHICS_ELEM_END) + GRAPHICS_ELEM_END.length();
             if (end > begin) {
+                String originalGraphElem = xmlDesc.substring(begin, end);
                 String graphElem = xmlDesc.substring(begin, end);
                 graphElem = graphElem.replaceAll("listen='[a-zA-Z0-9\\.]*'", "listen='" + target + "'");
                 graphElem = graphElem.replaceAll("address='[a-zA-Z0-9\\.]*'", "address='" + target + "'");
                 xmlDesc = xmlDesc.replaceAll(GRAPHICS_ELEM_START + CONTENTS_WILDCARD + GRAPHICS_ELEM_END, graphElem);
+                s_logger.debug(String.format("Replaced the VNC IP address [%s] with [%s] in VM [%s].", originalGraphElem, graphElem, vmName));
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java:473]:
```suggestion
            if (s_logger.isDebugEnabled()) {
                s_logger.debug(String.format("Replaced the VNC IP address [%s] with [%s] in VM [%s].", originalGraphElem, graphElem, vmName));
            }
```

**Comment by rohityadavcloud** [comment]
Can you merge latest main branch to your PR branch @SadiJr  ?

**Review by GabrielBrascher — APPROVED** [review]
Interesting log enhancements, LGTM.

**Pull request merged**